src/google/protobuf/compiler/cpp/cpp_message.cc

```diff
@@ -1121,12 +1122,12 @@ GenerateClassDefinition(io::Printer* printer) {
   if (HasGeneratedMethods(descriptor_->file(), options_)) {
     if (HasDescriptorMethods(descriptor_->file(), options_)) {
       printer->Print(vars,
-        "void CopyFrom(const ::google::protobuf::Message& from) PROTOBUF_FINAL;\n"
-        "void MergeFrom(const ::google::protobuf::Message& from) PROTOBUF_FINAL;\n");
+        "void CopyFrom(const ::google::protobuf::Message& from) PROTOBUF_FINAL PROTOBUF_OVERRIDE;\n"
```

Comment:
From what I understand you don't need "override" if you already have "final". Doesn't "final" imply "override" already?

Reply:
No. "override" and "final" have independent behavior. "final" means the function cannot be overridden, however, "override" means the function is overriding a base class function. If a function has both "override" and "final" it means, the function is overriding a base class function but cannot be overriden.

Reply:
I believe "final" has the exactly same semantics as "final override", i.e., it's overriding a virtual method in a base class and it can no longer be overridden.  In any case, please follow our C++ style guide:
https://google.github.io/styleguide/cppguide.html#Inheritance

Specifically:
```
Explicitly annotate overrides of virtual functions or virtual destructors with exactly one of an override or (less frequently) final specifier
```

Reply:
we use gcc "Wsuggest-override" warning option for compile time enforcement and that doesn't take "final" as replacement of "override". clang has better option called "Winconsistent-missing-override", but unfortunately not available on gcc yet.

Reply:
I think we should follow the style guide here. I just checked Google's internal code base and there is pretty much 0 usage of "final override" and "override final". Using it in protobuf code would require an exception granted by the style guide arbiters but I don't think it's justified for this case. Eventually gcc will have the same option, won't it?